### ISSUE TITLE
Eval bug: Cross-request KV-cache contamination on integrated GPUs: missing sync in process_ubatch races zero-copy graph inputs against in-flight compute

### ISSUE THREAD

**Issue by allenzz-dev** [ggml-org/llama.cpp#28056]
## Summary

On integrated GPUs (tested: AMD gfx1151 / Strix Halo APU, ROCm/HIP backend), `llama-server` can leak one request's conversation content into a later, unrelated request's responses. The KV cache gets corrupted during chunked prefill: K/V rows are scattered to wrong cells, and stale K/V from a previous request remains resident under cells the new request's attention mask marks visible. We hit this in production as verbatim fragments of a private chat appearing in another client's document-QA outputs.

Verified at v0.3.0 (build 10622, commit c1d0e7a00), HIP build, `-ngl 99 -b 2048 -ub 2048 -fa on -np 1`.

## Root cause (three links)

1. **Zero-copy graph inputs on iGPUs**: `ggml_backend_cuda_device_supports_buft` returns true for pinned-host bufts when `integrated` is set, so GPU kernels read graph-input tensors (`k_idxs`, `v_idxs`, `kq_mask`, `inp_tokens`, `inp_pos`, …) **directly from host memory**. On discrete GPUs these are split inputs with a protective synchronize+copy inside the sched — which is why this never reproduces on dGPUs.

2. **Synchronization gated on `pipeline_parallel`**: in `llama_context::process_ubatch`, the sync before `set_inputs` on the graph-reuse path only runs `if (cparams.pipeline_parallel)` — always false on a single GPU. The graph-rebuild path (`ggml_backend_sched_reset` + `ggml_backend_sched_alloc_graph`, which remap the compute buffers) has no synchronization at all.

3. **Fire-and-forget prefill**: `graph_compute` ends with `ggml_backend_sched_graph_compute_async`, and prefill chunks produce no outputs, so `llama_decode` returns without any sync and the server immediately submits the next chunk. Measured on this hardware: **six 2048-token chunks submitted within 60–70 ms while each takes 0.8–1.8 s to execute** (~6-deep async pipeline).

Result: ubatch N+1's `set_inputs` overwrites the very host buffers ubatch N's in-flight kernels are still reading. `ggml_set_rows` then scatters K/V rows to wrong cells; the intended cells keep stale data from a previous request, and the new request's mask marks them visible → cross-request contamination.

## Reproducer / evidence

Model: Qwen3.6-35B-A3B (hybrid GDN) Q6_K_XL, `llama-server --ctx-size 262144 -b 2048 -ub 2048 -fa on -np 1 --cache-reuse 1`, one client at a time, deterministic sampling (`temperature 0, top_k 1`) for the probe requests.

Scenario: 2-turn chat about a distinctive topic X → idle → send a ~21k-token synthetic document + 4 questions. The answers quote topic X verbatim (4/4 runs, up to 5 mentions of X's key term per answer).

Instrumentation findings (per-layer FNV-1a hash of individual K-cache rows, logged at every `seq_rm`):

- A probe cell inside the document's chunk-2 target range holds a **non-boot value planted during the chat phase** — which never legitimately writes anywhere near that cell — and that value **never changes while the document's 10 chunks are prefilled**, despite host-side `k_idxs` provably containing the correct contiguous indices for those chunks. So the same race has two faces: earlier writes scatter *into* wrong cells, and later writes *fail to land* on their intended cells.
- A control cell beyond the reach of both requests keeps its boot-time value throughout.
- Host-zeroing the attention K/V buffers between the two requests eliminates the contamination; zeroing the recurrent (GDN) state does not — the carrier is stale attention K/V, not the recurrent state.
- `GGML_CUDA_DISABLE_FUSION=1` reduces but does not eliminate it (slower enqueue → shallower pipeline — a timing effect, not a cause).

## Fix (verified)

Make the synchronize in `process_ubatch` unconditional, hoisted above the `can_reuse` branch so it also covers the rebuild path:

```cpp
    // the previous graph_compute_async may still be running on the GPU. synchronize before
    // mutating any graph state: set_inputs overwrites input tensors the in-flight compute
    // may still be reading, and sched_reset/alloc_graph remap the compute buffers under it.
    // not limited to pipeline parallelism: on integrated GPUs graph inputs are read
    // zero-copy from pinned host memory.
    ggml_backend_sched_synchronize(sched.get());

    if (!graph_reuse_disable && res->can_reuse(gparams)) {
        n_reused++;
    } else {
        ...
```

After the fix: 0/12 contaminated outputs across 3 runs of the reproducer (was 4/4 heavy per run), and the K-cell hash trajectory becomes healthy (chat era untouched, document data lands in its cells and stays there through all 4 questions).

**Performance cost is zero on this hardware** — the async pipeline only queued deeper, the GPU was already saturated:

| metric | before | after |
|---|---|---|
| 21k-token prefill | 1230.8 t/s | 1231.8 t/s |
| 2k-token prefill | 957–970 t/s | 957–966 t/s |
| decode | 48.5 t/s | 48.4 t/s |

## Notes

- Affects any platform where graph inputs live in host-visible buffers read zero-copy by the device; the trigger is any run of ubatches with no output reads in between (typical: chunked prefill of a large prompt). Decode is unaffected because per-step logits reads already synchronize.
- A conservative alternative would gate the sync on the integrated/zero-copy condition, but the unconditional sync measured free here and is much safer.
- Possibly related: #27612 (gfx1151 ROCm tool-call flakiness where Vulkan works — consistent with occasional prefill KV corruption on the same APU family).

Happy to submit this as a PR.

**Comment by machdat** [comment]
Independent confirmation on the same hardware class, with a different model and a much more visible symptom: the corruption also hits **within a single request**, no cross-request scenario needed.

**Environment:** AMD Ryzen AI MAX+ 395 (Strix Halo, gfx1151), ROCm 7.2.4, llama.cpp `c060ca9`, HIP build (`GGML_HIP=ON`, `GGML_HIP_GRAPHS=ON`, `GGML_HIP_MMQ_MFMA=ON`). Model: GLM-4.7-Flash (arch `deepseek2`), unsloth GGUF, sha256 verified against the LFS entry.

**Symptom:** with a ~1000-token prompt, output is word salad from the first token — repeated fragments like ` ```bash EOF ``` ` over and over, never terminating. With a 20-token prompt, the same model on the same GPU is perfectly coherent.

**The trigger is exactly the one you name — a prefill split across several ubatches.** All runs below use `llama-cli` directly (no server, no agent harness), `--temp 0 --seed 1 -st --jinja -ngl 999 -c 8192`, identical prompt, only `-ub` varies:

| `-ub` | result |
|---|---|
| 512 (default) | word salad |
| 640 | word salad |
| 768 | coherent but off-task |
| 896 | correct |
| 1024 | correct |

So it degrades gradually with the number of splits rather than flipping at a threshold. Doubling the prompt (~2000 tokens) moves the boundary with it: `-ub 4096` correct, `-ub 1024` word salad. It is the splitting, not the length.

**Further isolation:**

- **Vulkan build, same commit, same prompt, `-ub 512`: correct.** ROCm/HIP only.
- **Qwen3.6-35B-A3B, same prompt, `-ub 512`: correct.** So it is not every model — worth noting since your reproducer used Qwen3.6-35B; our Qwen run had only 2 chunks, yours 6–10, which may explain the difference.
- `-c 4096 / 32768 / 65536 / 131072`: no difference.
- Q4_K_XL and Q6_K_XL: identical behaviour. `--flash-attn off`: identical.
- `--n-cpu-moe 99` (all expert FFNs on CPU): **still corrupt** — consistent with the carrier being attention K/V rather than anything in the expert path.
- CPU-only (`-ngl 0`): correct.

**Minimal reproducer:**

```
llama-cli -m GLM-4.7-Flash-UD-Q4_K_XL.gguf \
    -sysf system.txt -f user.txt \
    -n 100 --temp 0 --seed 1 -st --jinja -ngl 999 -c 8192 -ub 512 -b 2048
```

with any system+user prompt totalling ~1000 tokens. Swapping `-ub 512` for `-ub 2048` (single ubatch) makes it correct, deterministically, at temperature 0.

**Practical workaround** for anyone hitting this before the fix lands: raise `--ubatch-size`/`--batch-size` so a typical prompt fits into one ubatch. In our agent workload (mini-swe-agent) that turns a completely unusable model into one that solves the task in 75 s — first-turn prefill fits in one ubatch, and subsequent turns only prefill small deltas thanks to the server prompt cache. It is a workaround, not a fix: whenever the cache misses and a large chunk is prefilled at once, the corruption can return mid-run.

**Your patch fixes it here too.** Applied to `c060ca9` (unconditional `ggml_backend_sched_synchronize` hoisted above the `can_reuse` branch, exactly as written in your report), rebuilt, and re-ran the cases that were reliably broken before:

| case | before | after |
|---|---|---|
| ~1000-token prompt, `-ub 512` | word salad | correct, on-task |
| same, `-ub 640` | word salad | correct |
| same, `-ub 128` (many splits) | word salad | correct |
| ~2000-token prompt, `-ub 1024` | word salad | correct |

So the fix holds for a second model family (`deepseek2`/MLA) and for the intra-request symptom, not just for the cross-request contamination you documented.

One honest limit on our data: the boundary is not fully mapped — server-side runs with 775- and 563-token prefills at `-ub 512` were coherent, so a split alone is not sufficient; the size of the trailing chunk appears to matter too.

**Comment by afq984** [comment]
Independently found reproducer on Strix Halo (gfx1151, ROCm 7.2.4, HIP backend), build 10566 / bb4caa7540, with hybrid VL models (Qwen3.5-4B UD-Q4_K_XL + mmproj, and Qwen3.8-27B).

Server: llama-server -m Qwen3.5-4B-UD-Q4_K_XL.gguf --mmproj mmproj-F16.gguf -ngl 99 -c 16384 --jinja --reasoning off --parallel 1 --cache-ram 0

I also confirmed that #28058 fixes the reproducer.

```python
#!/usr/bin/env python3
"""Minimal repro: cross-conversation state leak in llama-server with a
hybrid-recurrent Qwen3.5-class VL model (recurrent/linear-attention layers).

Three requests against a llama-server started with:

  llama-server -m Qwen3.5-4B-UD-Q4_K_XL.gguf --mmproj mmproj-F16.gguf \
    -ngl 99 -c 32768 --jinja --reasoning off --parallel 1 --cache-ram 0 \
    --host 127.0.0.1 --port 8080   # add -dev CUDA0 / Vulkan0 / none to pick a backend

  P1: probe request containing ONE image, temperature 0, fixed seed
  X : an unrelated text-only conversation (unique prefix -> zero overlap)
  P2: the probe again, byte-compared with P1

Expected: P1 == P2. Observed: P2 differs and reproduces conversation X's
content (proper nouns, scenery) inside the probe persona's reply.
Text-only probes (no image) are NOT affected; the affected request must
contain at least one image. --cache-ram 0 rules out the host prompt cache.

Only dependency: pip install openai. The image is generated in-script.
"""
import base64
import struct
import sys
import uuid
import zlib

from openai import OpenAI

import os
BASE_URL = os.environ.get("LLAMA_URL", "http://127.0.0.1:8080/v1")
API_KEY = os.environ.get("LLAMA_API_KEY", "none")

W, H = 256, 256  # flat gray PNG, stdlib-generated


def flat_png(w, h, rgb=(128, 128, 128)):
    def chunk(tag, data):
        c = tag + data
        return struct.pack(">I", len(data)) + c + struct.pack(">I", zlib.crc32(c))
    ihdr = struct.pack(">IIBBBBB", w, h, 8, 2, 0, 0, 0)
    row = b"\x00" + bytes(rgb) * w
    idat = zlib.compress(row * h)
    return (b"\x89PNG\r\n\x1a\n" + chunk(b"IHDR", ihdr)
            + chunk(b"IDAT", idat) + chunk(b"IEND", b""))


IMG_URL = "data:image/png;base64," + base64.b64encode(flat_png(W, H)).decode()

client = OpenAI(base_url=BASE_URL, api_key=API_KEY, timeout=600)


def chat(system, content, max_tokens, temperature=0.0):
    r = client.chat.completions.create(
        model=os.environ.get("LLAMA_MODEL", "default"),
        messages=[{"role": "system", "content": system},
                  {"role": "user", "content": content}],
        temperature=temperature, top_p=1.0, max_tokens=max_tokens,
        extra_body={"seed": 42},
    )
    return r.choices[0].message.content or ""


PROBE_SYS = ("You are Runa, warden of the Winter Gate in the frozen realm of "
             "Frosthaven. Your wolf Ylva walks beside you. The attached image "
             "is one of your memories. Stay in character.")
PROBE_CONTENT = [
    {"type": "image_url", "image_url": {"url": IMG_URL}},
    {"type": "text", "text": "Describe your morning in about 150 words."},
]
OTHER_SYS = ("You are Ser Kaelen, Flamewarden of the Obsidian Citadel in the "
             "burning realm of Emberfall. Your dragon Sindra carries you. "
             "Stay in character.")
OTHER_MSG = ("Tell the story of your day at length: your world, your dragon, "
             "your rival Lord Ashten of the Cinder Court, and the Phoenix "
             "Festival.")

if "--probe-only" in sys.argv:
    # control: run only the probe (for byte-comparison across server restarts)
    print(chat(PROBE_SYS, PROBE_CONTENT, 250))
    sys.exit(0)

print("P1 ...", flush=True)
p1 = chat(PROBE_SYS, PROBE_CONTENT, 250)
print("X  ...", flush=True)
chat(f"[fixed-prefix] {OTHER_SYS}", OTHER_MSG, 300, temperature=0.85)  # fixed prefix => deterministic
print("P2 ...", flush=True)
p2 = chat(PROBE_SYS, PROBE_CONTENT, 250)

print("=" * 70)
if p1 == p2:
    print("IDENTICAL - no leak in this configuration")
    sys.exit(0)
import re
leaked = [w for w in ("emberfall", "sindra", "dragon", "obsidian", "citadel", "phoenix",
                     "ashten", "cinder", "flamewarden", "kaelen")
          if re.search(r"\b" + w + r"\b", p2, re.I) and not re.search(r"\b" + w + r"\b", p1, re.I)]
print("DIFFERENT - conversation X influenced the probe. Terms from X's PROMPT now in P2:", leaked or "none")
import difflib
print("\n".join(list(difflib.unified_diff(
    p1.splitlines(), p2.splitlines(), "P1", "P2", lineterm=""))[:40]))
sys.exit(1)
```
